test/support/utils.rb

```diff
@@ -14,39 +14,67 @@ def create_message fail_msg
       "#{fail_msg}\n[DEBUG SESSION LOG]\n> " + @backlog.join('> ')
     end
 
-
     def combine_regexps(regexps)
       Regexp.new(regexps.map(&:source).reduce(:+))
     end
 
     def debug_code(program, **options, &block)
       @queue = Queue.new
-      block.call
+      @scenario = block
+      @scenario.call
       write_temp_file(strip_line_num(program))
-      create_pseudo_terminal(**options)
+      setup_terminal(**options)
       check_line_num!(program)
     end
 
     DEBUG_MODE = false
+    ASK_CMD = %w[quit delete kill]
 
     def debug_print msg
       print msg if DEBUG_MODE
     end
 
     RUBY = RbConfig.ruby
-    REPL_RPOMPT = /\(rdbg\)/
+    RUBY_DEBUG_TEST_PORT = '12345'
 
-    def create_pseudo_terminal(boot_options: "-r debug/run")
+    def setup_terminal(boot_options: "-r debug/run", remote: true)
+      ENV['RUBY_DEBUG_TEST_LOCAL_ONLY_MODE'] = 'false' # for fast debug test
       inject_lib_to_load_path
+
+      if remote && ENV['RUBY_DEBUG_TEST_LOCAL_ONLY_MODE'] == 'false'
+        setup_terminal(remote: false) # local test will be executed first
+        @mutex = Mutex.new
+        boot_options = "-r debug/open"
+        repl_prompt = /\(rdb\)/
```

Comment:
it'd be great if we can add some comments to this section to make things clearer

```ruby
        # run test in local mode
        cmd = "#{__dir__}/../../exe/rdbg -A"
        new_child_process("#{RUBY} #{boot_options} #{temp_file_path}")

        # run test in remote mode
        repl_prompt = /\(rdb\)/
        create_pseudo_terminal(cmd, repl_prompt)
        cmd = "#{__dir__}/../../exe/rdbg -A #{RUBY_DEBUG_TEST_PORT}"
        new_child_process("#{__dir__}/../../exe/rdbg -O --port=#{RUBY_DEBUG_TEST_PORT} #{temp_file_path}")
```

Reply:
https://github.com/ruby/debug/pull/76/files#diff-ee14bf65f14927f7778a246f90c5cec0c581be3e2658c2a8a3726c59db5be3f9R50-R62